### PR TITLE
lsp: Store diagnostics for unloaded buffers

### PR DESCRIPTION
**Pull request by mfussenegger**
To avoid loading buffers https://github.com/neovim/neovim/pull/12440
changed the logic to not process diagnostics for unloaded buffers.

This is problematic for language servers where compile errors or build
errors are reported via diagnostics. These errors may prevent the
language server from providing all functions and it is difficult for
users to debug it without having access to the errors.

For example, with eclipse.jdt.ls there may be a problem with gradle (the
build tool for java), it results in a diagnostics like this:

    org.gradle.toolingapi/build.gradle|1 col 1| Could not run build action using Gradle distribution 'https://services.gradle.org/distributions/gradle-4.8.1-bin.zip'.

This would be invisible to users unless the user happens to open the
right file. In this case the user would actually never see the error,
because the language server isn't attached to the build configuration
files.

This changes the behaviour to at least store the diagnostics. The other
operations which are more expensive are still skipped.

### CODE REVIEW

File: runtime/lua/vim/lsp/callbacks.lua [hunk 1]
```diff
@@ -104,6 +92,23 @@ M['textDocument/publishDiagnostics'] = function(_, _, result)
...
+  -- unloaded buffers.
+  if not api.nvim_buf_is_loaded(bufnr) then
+    -- Still save the diagnostics
+    -- Language servers may report compile or build errors via diagnostics
+    -- Users should be able to find these, even if they're in files which
+    -- are not loaded.
+    util.buf_diagnostics_save_positions(bufnr, result.diagnostics)
+    return
+  end
+
```
Comment (tjdevries) [inline runtime/lua/vim/lsp/callbacks.lua:111]: We now save the diagnostics twice?

Comment (mfussenegger) [reply runtime/lua/vim/lsp/callbacks.lua:111]: It returns if the buffer is not loaded and if the buffer is loaded the `save_positions` call further above doesn't happen.

**Comment by mfussenegger** [comment]
Another case where users are unlikely to ever have the buffer loaded is for diagnostics on folders, which can also happen.


This is an example error when trying to use eclipse.jdt.ls on the Elasticsearch code base:

```
x-pack/plugin/async|1 col 1| One or more cycles were detected in the build path of project ':x-pack:plugin:async'. The paths towards the cycle and cycle are: :x-pack:plugin:async->{:libs:elasticsearch-secure-sm, :test:framework, :libs:elasticsearch-nio, :server, :libs:elasticsearch-x-content, :libs:elasticsearch-cli, :libs:elasticsearch-core} :x-pack:plugin:async->{:libs:elasticsearch-secure-sm, :test:framework, :libs:elasticsearch-nio, :server, :libs:elasticsearch-x-content, :libs:elasticsearch-cli, :libs:elasticsearch-core, :libs:elasticsearch-geo} :x-pack:plugin:async->{:client:rest-high-level, :x-pack:plugin:eql, :x-pack:plugin:security, :x-pack:plugin:monitoring, :x-pack:plugin:watcher, :x-pack:plugin:core}
```


My plan would be to pick-up https://github.com/neovim/neovim/pull/12578 after this is merged and extend it to also output diagnostic errors if present (maybe limited to 5 or 10 or so).

Otherwise it is hard for users to discover issues like that.

File: runtime/lua/vim/lsp/callbacks.lua (hunk 1 above)
Comment (tjdevries) [reply runtime/lua/vim/lsp/callbacks.lua:111]: Ah, yes, that's obviously correct. Could we instead just pull the shared `util.buf_diagnostics_save_positions(bufnr, result.diagnostics)` above this `if` block and remove it from line 112? I'd prefer to not have it copied between the two places, if that makes sense.

**Comment by tjdevries** [comment]
Perfect, I'll wait for green checkmarks and then we will merge. Thanks!

**Comment by mfussenegger** [comment]
@tjdevries all tests turned green. Is this good to merge?

**Pull request merged**